backend/internal/service/export_service.go

```diff
@@ -0,0 +1,217 @@
+package service
+
+import (
+	"archive/zip"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+	"unicode/utf8"
+
+	"github.com/pocket-id/pocket-id/backend/internal/common"
+	"gorm.io/gorm"
+)
+
+// ExportService handles exporting Pocket ID data into a ZIP archive.
+type ExportService struct {
+	db *gorm.DB
+}
+
+func NewExportService(db *gorm.DB) *ExportService {
+	return &ExportService{db: db}
+}
+
+// ExportToZip performs the full export process and writes the ZIP data to the given writer.
+func (s *ExportService) ExportToZip(w io.Writer) error {
+	dbData, err := s.extractDatabase()
+	if err != nil {
+		return err
+	}
+
+	return writeExportZipStream(w, dbData)
+}
+
+// extractDatabase reads all tables into a DatabaseExport struct
+func (s *ExportService) extractDatabase() (DatabaseExport, error) {
+	tables, err := s.listTables()
+	if err != nil {
+		return DatabaseExport{}, err
+	}
+
+	version, err := s.schemaVersion()
+	if err != nil {
+		return DatabaseExport{}, err
+	}
+
+	out := DatabaseExport{
+		Provider: s.db.Name(),
+		Version:  version,
+		Tables:   map[string][]map[string]any{},
+	}
+
+	for _, table := range tables {
+		if err := s.dumpTable(table, &out); err != nil {
+			return DatabaseExport{}, err
+		}
+	}
+
+	return out, nil
+}
+
+func (s *ExportService) listTables() ([]string, error) {
+	var tables []string
+
+	switch common.EnvConfig.DbProvider {
+	case common.DbProviderSqlite:
+		if err := s.db.Raw(`SELECT name FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%' AND name != 'schema_migrations';`).Scan(&tables).Error; err != nil {
+			return nil, fmt.Errorf("failed to query sqlite tables: %w", err)
+		}
+
+	case common.DbProviderPostgres:
+		if err := s.db.Raw(`SELECT tablename FROM pg_tables WHERE schemaname='public' AND tablename != 'schema_migrations';`).Scan(&tables).Error; err != nil {
```

Comment:
Nit, should probably make the schema of pg_tables explicit just to be extra safe
```suggestion
		if err := s.db.Raw(`SELECT tablename FROM pg_catalog.pg_tables WHERE schemaname='public' AND tablename != 'schema_migrations';`).Scan(&tables).Error; err != nil {
```